homework5/ParsingTree/ParsingTree/Tree.cs

```diff
@@ -0,0 +1,138 @@
+    /// </summary>
+    /// <param name="root">The tree root.</param>
+    public Tree(Node root)
+    {
+        this.root = root;
+    }
+
+    /// <summary>
+    /// The function for calculating the value.
+    /// </summary>
+    /// <returns>The result of calculating.</returns>
+    public int Evaluate()
+    {
+        if (this.root == null)
+        {
+            System.Console.WriteLine("The tree must not be empty.");
+            return -1;
+        }
+
+        return this.root.Evaluate();
+    }
+
+    /// <summary>
+    /// The function for output to the console.
+    /// </summary>
+    /// <returns>String.</returns>
+    public string Print()
+    {
+        if (this.root == null)
+        {
+            throw new ArgumentNullException();
+        }
+
+        return this.root.Print();
+    }
+
+    /// <summary>
+    /// Looking for a balance of brackets.
+    /// </summary>
+    /// <param name="operands">The expression.</param>
+    /// <returns>Position of split space.</returns>
+    public int FindOperandSplitPos(string operands)
+    {
+        int balance = 0;
+        for (int i = 0; i < operands.Length; ++i)
+        {
+            char c = operands[i];
+            if (c == '(')
+            {
+                ++balance;
+            }
+            else if (c == ')')
+            {
+                --balance;
+            }
+
+            if (c == ' ' && balance == 0)
+            {
+                return i;
+            }
+        }
+
+        return -1;
+    }
+
+    /// <summary>
+    /// Builds a parse tree.
+    /// </summary>
+    /// <param name="data">Started expression.</param>
+    /// <returns>The root of tree.</returns>
+    /// <exception cref="FormatException">If expression must be enclosed in parentheses.</exception>
+    /// <exception cref="InvalidOperationException">If you select an operation that does not exist.</exception>
+    public Node GetParsingTree(string data)
+    {
+        data = data.Trim();
+        if (int.TryParse(data, out int number))
+        {
+            return new NumberNode(number);
+        }
+
+        if (!data.StartsWith("(") || !data.EndsWith(")"))
+        {
+            throw new FormatException("Expression must be enclosed in parentheses");
+        }
+
+        string inner = data.Substring(1, data.Length - 2).Trim();
+        int firstSpace = inner.IndexOf(' ');
+        if (firstSpace <= 0)
+        {
+            throw new FormatException("Invalid expression format");
+        }
+
+        string operation = inner.Substring(0, firstSpace);
+        string operands = inner.Substring(firstSpace + 1).Trim();
+        int splitPos = this.FindOperandSplitPos(operands);
+        if (splitPos < 0)
+        {
+            throw new FormatException("Could not split operands");
+        }
+
+        string leftPart = operands.Substring(0, splitPos).Trim();
+        string rightPart = operands.Substring(splitPos).Trim();
```

Comment:
Начиная с C# 8, можно использовать операторы диапазона. То есть
```suggestion
        string leftPart = operands.[..splitPos].Trim();
        string rightPart = operands.[splitPos..].Trim();
```